smt.go

```diff
@@ -547,107 +526,112 @@ func (smt *SMT) ProveClosest(path []byte) (
-		if err != nil {
-			return
-		}
-		return &ext, nil
+// resolveNode returns a trieNode (inner, leaf, or extension) based on what they
+// keyHash points to.
+func (smt *SMT) resolveNode(digest []byte) (trieNode, error) {
+	// Check if the keyHash is the empty zero value of an empty subtree
+	if bytes.Equal(smt.placeholder(), digest) {
+		return nil, nil
 	}
-	leftHash, rightHash := smt.th.parseNode(data)
-	inner := innerNode{persisted: true, digest: hash}
-	inner.leftChild, err = resolver(leftHash)
+
+	// Retrieve the encoded noe data
+	data, err := smt.nodes.Get(digest)
 	if err != nil {
-		return
+		return nil, err
 	}
-	inner.rightChild, err = resolver(rightHash)
-	if err != nil {
-		return
+
+	// Return the appropriate node type based on the first byte of the data
+	if isLeafNode(data) {
+		path, valueHash := smt.parseLeafNode(data)
+		return &leafNode{
+			path:      path,
+			valueHash: valueHash,
+			persisted: true,
+			digest:    digest,
+		}, nil
+	} else if isExtNode(data) {
+		pathBounds, path, childData := smt.parseExtNode(data)
+		return &extensionNode{
+			path:       path,
+			pathBounds: [2]byte(pathBounds),
+			child:      &lazyNode{childData},
+			persisted:  true,
+			digest:     digest,
+		}, nil
+	} else if isInnerNode(data) {
+		leftData, rightData := smt.th.parseInnerNode(data)
+		return &innerNode{
+			leftChild:  &lazyNode{leftData},
+			rightChild: &lazyNode{rightData},
+			persisted:  true,
+			digest:     digest,
+		}, nil
+	} else {
+		panic("invalid node type")
 	}
-	return &inner, nil
 }
 
-func (smt *SMT) resolveSum(hash []byte, resolver func([]byte) (trieNode, error),
-) (ret trieNode, err error) {
-	if bytes.Equal(placeholder(smt.Spec()), hash) {
-		return
+// resolveNode returns a trieNode (inner, leaf, or extension) based on what they
+// keyHash points to.
+func (smt *SMT) resolveSumNode(digest []byte) (trieNode, error) {
+	// Check if the keyHash is the empty zero value of an empty subtree
+	if bytes.Equal(smt.placeholder(), digest) {
+		return nil, nil
 	}
-	data, err := smt.nodes.Get(hash)
+
+	// Retrieve the encoded noe data
+	data, err := smt.nodes.Get(digest)
 	if err != nil {
 		return nil, err
 	}
-	if isLeaf(data) {
-		leaf := leafNode{persisted: true, digest: hash}
-		leaf.path, leaf.valueHash = parseLeaf(data, smt.ph)
-		return &leaf, nil
-	}
-	if isExtension(data) {
-		ext := extensionNode{persisted: true, digest: hash}
-		pathBounds, path, childHash, _ := parseSumExtension(data, smt.ph)
-		ext.path = path
-		copy(ext.pathBounds[:], pathBounds)
-		ext.child, err = resolver(childHash)
-		if err != nil {
-			return
-		}
-		return &ext, nil
-	}
-	leftHash, rightHash := smt.th.parseSumNode(data)
-	inner := innerNode{persisted: true, digest: hash}
-	inner.leftChild, err = resolver(leftHash)
-	if err != nil {
-		return
-	}
-	inner.rightChild, err = resolver(rightHash)
-	if err != nil {
-		return
+
+	// Return the appropriate node type based on the first byte of the data
+	if isLeafNode(data) {
```

Comment:
I feel like we can introduce a helper as this logic is duplicated

Reply:
Refactored to use helpers but the logic is unique between a sum & non-sum.